tests/binary_agreement.rs

```diff
@@ -27,31 +27,139 @@ extern crate threshold_crypto as crypto;
+        if self.current_epoch < msg.message.epoch {
+            self.current_epoch = msg.message.epoch;
+            // Reset for the next epoch.
+            self.sent_attack = false;
+        }
+    }
+
+    fn step(&mut self) -> Vec<MessageWithSender<BinaryAgreement<NodeId>>> {
+        let mut messages = Vec::new();
+        if !self.sent_attack {
+            self.sent_attack = true;
+            for &our_node_id in &self.known_adversarial_ids {
+                let f = self.known_node_ids.len() / 3;
+                for &target in &self.known_node_ids[0..f] {
+                    // Disagree with the first third.
+                    messages.push(MessageWithSender::new(
+                        our_node_id,
+                        TargetedMessage {
+                            target: Target::Node(target),
+                            message: Message {
+                                epoch: self.current_epoch,
+                                content: MessageContent::SbvBroadcast(
+                                    sbv_broadcast::Message::BVal(true),
+                                ),
+                            },
+                        },
+                    ));
+                }
+                for &target in &self.known_node_ids[f..(f * 2)] {
+                    // Affirm the second third.
+                    messages.push(MessageWithSender::new(
+                        our_node_id,
+                        TargetedMessage {
+                            target: Target::Node(target),
+                            message: Message {
+                                epoch: self.current_epoch,
+                                content: MessageContent::SbvBroadcast(
+                                    sbv_broadcast::Message::BVal(false),
+                                ),
+                            },
+                        },
+                    ));
+                }
+            }
+        }
+        messages
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum InputConfig {
+    /// Always input this bool.
+    Fixed(bool),
+    /// Input a random bool.
+    Random,
+    /// Deterministically split the network.
+    /// Input false to the first 2/3, and true to the rest.
+    /// This is used to test the reordering attack.
+    Split,
+}
 
 fn test_binary_agreement<A: Adversary<BinaryAgreement<NodeId>>>(
     mut network: TestNetwork<A, BinaryAgreement<NodeId>>,
-    input: Option<bool>,
+    input: InputConfig,
 ) {
+    // The borrow checker gets in our way here, so we clone the node ids.
     let ids: Vec<NodeId> = network.nodes.keys().cloned().collect();
-    for id in ids {
-        network.input(id, input.unwrap_or_else(rand::random));
+    for (i, id) in ids.into_iter().enumerate() {
+        let value = match input {
+            InputConfig::Fixed(b) => b,
+            InputConfig::Random => rand::random(),
+            InputConfig::Split => i < (2 * network.nodes.len() / 3),
```

Comment:
This will only test constant input: `true` for the set `A` and `false` for the set `B`. There can be a second attack with these values negated. That possible attack is not covered here.